### DISCUSSION TITLE
Day 2 lecture & workshop outline

### DISCUSSION THREAD

**Discussion by matchings** [AllenSWDB/SWDB_2024_Physiology#2] [Ideas]
**Version 1 lecture outline proposal for day 2 - single cell & population coding** 
 
Background & motivation

Classical theories of coding - often looking for a representation of something predetermined (ex: a sensory stimulus), often one cell at a time, mostly excitatory cells
  - Methods, model systems & theories people use to look at this
     - PSTH, tuning curves, rate vs. timing codes
     - Retina, early sensory systems, anesthetized animals
   - Examples from the literature
     - Hubel & Wiesel - basic sensory tuning
     - Barlow, Attneave, Rieke - efficient coding
     - Olshausen & Field - sparse coding 
 
--> Research over the last few decades has shown us that neural activity across the brain (even in sensory areas) encodes behavior and task variables, is modulated by context, learning, etc. 
--> With the advent of large scale recording methods, we can now think about coding and computation by populations of neurons 
--> We have also discovered a huge amount of cell type diversity 
--> However we still don’t have a handle on what / when / where / how/ by whom things are encoded or decoded
--> What methods & data can we use to 
 
Updated theories of coding - multiplexed coding, state dependent coding, experience dependent coding, population level computation
  - Methods & model systems & theories people use to look at this
    - ML methods, GLMs, classification / decoding, neural networks 
    - Multi-area / multi-layer recordings in behaving animals 
  - Examples from the literature
    - Behavior modulation, context dependence, learning
      - Niell, Stringer, Churchland, etc - activity modulated by running / behavior state
      - Kuchibhotla et al - task engagement alters coding
      - Ni & Cohen - correlations with attention & learning
    - Different types of behaviors
      - Juavinett & reviews on axes of behavior / task types 
      - Head fixed, freely moving, constrained tasks, natural behaviors
    - Cell types & coding
      - Vip cells & running & behavior & omissions & novelty
      - Coding in more refined cell types - Condylis, Bugeon

**Comment by matchings** [comment]
**Outline of 2023 workshops**
 
Single cell coding Workshop 1
1.	Differences between ophys & ephys
2.	Intro to VCO dataset
3.	Direction tuning
4.	Reliability of responses
 
Single cell coding Workshop 2
1.	Regression
a.	Fitting a first order polynomial
b.	Linear regression
c.	Cross validation
d.	Modeling activity
e.	Computing tuning curves as a regression problem
f.	Multiple linear regression
 
Population Coding Workshop 1
1.	VCN dataset overview
2.	Firing rates
3.	Population activity
4.	Classifier on population data 
a.	SVM
b.	Random forest
 
Population Coding Workshop 2 
1.	Correlations 
2.	PCA
3.	Signal & noise correlations 
4.	Decoding with & without noise correlations

**Reply by matchings** [reply]
Note from discussion with Michael - much of last year's material on regression and fitting can go into a standalone notebook / tutorial in the databook focused on principles of regression, and our workshop for this year can move more quickly to implementing regression via scikit-learn

**Comment by matchings** [comment]
Workshop guidelines
 
•	Goal is to go from a question to a set of methods and appropriate datasets
•	"An important aspect of this approach is explaining both how the data is appropriate for the question (or might be) and also what steps you need to take to put the data in a form appropriate for the computational method"

**Comment by matchings** [comment]
**2024 workshops - Option 1**
 
Questions: How do single cells encode stimulus & behavioral information? How do populations of neurons encode stimulus & behavioral information?

Workshop 1 - single cell coding
- Tuning
- Regression
Workshop 2 - population coding
- Correlations
- Decoding

Option 1 - Examples of sensory coding & behavioral coding in each workshop
**Option 2 - Focus on either sensory coding / decoding or choice coding / decoding (and make the other one an exercise)**

Workshop 1 - single cell coding: 

1. What tools / methods / data do you need?
  - Considerations:
    - Fine timescale / PSTH - VBN
    - Cell type information - VBO
    - Simultaneous measurement of stimulus / behavior + neural activity
    - Coding for different sensory & behavioral features
    - Event based vs continuous information 
  -  Visual Behavior dataset (pick ophys or NP)
2. Tuning curves 
    - What is the average response to a stimulus? 
    - Tuning for running speed (trial averaged)
    - Exercises:
      - Trial to trial variability in sensory response or behavior
3. Regression 
    - How does activity track an external variable over time? 
        - Running speed, pupil
    - Splitting & validation
    - Exercises:
      - Multiple linear regression 
         - How do multiple features (stimulus + behavior) contribute to single cell activity? 
      - GLMs, other scikit-learn models you could swap for
 
Workshop 2 - population coding: 

1. Tools / methods / data 
    - Visual Behavior Neuropixels
      - Concurrent recordings across multiple areas & depths
      - Behavior measures (running & pupil) but also behavioral choices (hit/miss) and behavioral states (passive/active)
2. Population activity patterns
    - Rasters of activity 
    - Trial to trial variability in activity 
4. Signal & noise correlations 
    - Similarity of tuning across neurons
    - Shared trial to trial variability
    - Behavior state / choice influence on sensory responses
5. Classification / decoding 
    - Decoding stimulus identity
    - Decoding behavior state / behavioral choice

Exercises / problem sets: 
- Try with a different dataset (ophys vs NP)
- Try for pupil instead of running 
- Swap out for a different model
- Try for a different condition (active/passive, familiar/novel)
- Try for a different cre line / brain area 
- Try decoding behavioral response after omissions instead of changes

**Reply by matchings** [reply]
6/13/24 planning meeting

Questions: 
    - Are there any concepts or methods from the other days that they want us to highlight or emphasize?
    - Are we duplicating anything conceptually from other days? 

Notes: 
- Leave out behavior state (that's for day 4), focus on behavioral choice
  - Could go with VBO then since day 4 will do VBN?
- Keep it simple (MVP version), put elaborations in exercises
- Hint at exercises & project ideas throughout workshops
- Include David Liu in any meetings on materials development
- Capsules due July 1

**Comment by matchings** [comment]
**2024 workshops - Option 2**
 
Questions: How do single cells and populations of neurons encode sensory information? How do single cells & populations encode behavior & task information? How are sensory codes modified by behavior state & context? 
 
Workshop 1 - sensory coding
Workshop 2 - behavioral coding
_Examples of single cell & population coding in each workshop_

Workshop 1 - sensory coding: 

1. Recap of VCN/O dataset & suitability for the question
    - Different stimuli
    - Cre lines vs. simultaneous multi-area at high temporal resolution
2. Single cell coding for stimulus information
    - Single cell tuning curve
    - Reliability
    - Regression
      - Train test split / validation
      - Swapping out for a different model
3. Population coding for stimulus information
    - Visualizing population activity
    - Signal & noise correlations 
    - Classification / decoding
 
Workshop 2 - behavior coding: 

1. Recap of VBN/O dataset & suitability for the question
    - Behavioral task, different contexts
    - Active & passive blocks, familiar & novel stimuli 
    - Multi-area, multi-depth recordings, optotagging
2. Single cell coding
    - Tuning for behavior variables (running, pupil) 
    - Regression for behavior tuning 
    - Multiple linear regression - joint coding of stimulus & behavior
3. Population coding
    - State dependent stimulus tuning curves (run vs not run, active vs. passive)
    - Context dependent noise correlations (running vs not run, active vs. passive)
    - Decoding task variables (hit / miss)

**Comment by matchings** [comment]
Notes about incorporating feedback from 6/13 meeting

Main feedback - focus on behavior choice, not behavior state

One concern with this is that splitting trials by hit vs miss to look at behavioral choice is effectively sorting the data by behavioral state, because most misses occur during disengaged periods. The right thing to do for this analysis is to first split the data by engagement state, then compare hits and true perceptual misses. However this makes things much more complicated, so we can just ignore this fact for now and leave it for an exercise. 

One other point is that familiarity / novelty is a major part of this dataset, and there are big differences in coding between familiar & novel images. This can be incorporated by comparing across sessions for VBO, or across images within the same session for VBN.

**Comment by matchings** [comment]
**Final outline** 

Questions: How do single cells encode stimulus & behavioral information? How do populations of neurons encode stimulus & behavioral information? How does experience influence neural coding?

**Workshop 1 - single cell coding**
1. Dataset intro 
    - Visual Behavior Ophys dataset
    - Cre lines, session types, areas & depths
    - Select session to analyze - Sst cells, familiar sessions, V1 
    - Aligning traces to stimulus onset times (brain_observatory_utilities)
2. Tuning
    - Mean response to images
    - Variability in image response
   - What could cause this variability? 
    - Color trials by running vs not running, color trials by hit vs miss
3. Regression
    - Plot dFF vs run speed or dFF vs. choice
    - Linear regression
      - Fit the curve to look at tuning for running or choice
      - Train, test, split --> fit, predict
    - Multiple linear regression
      - Do we get better predictions with a combination of image & running or image & choice? 
 4. Impact of experience
     - Predict activity of the same cell in the novel session
 


**Workshop 2 - population coding**
1. Dataset
    - VBN: interleaved familiar & novel stimuli
    - Excitatory cells, multiple areas & depths
    - Spike times, population activity patterns
2. Decoding
    - Decoding stimulus identity
      - Does accuracy differ for familiar vs novel images? 
      - Does accuracy differ for hits vs. misses? 
    - Try for different time bins
    - Try for different #s of neurons
 3. Impact of correlations
    - How do correlations impact ability to decode?
    - Comparing image tuning across neurons - signal correlations
    - Comparing trial to trial variability across neurons - noise correlations
    - Signal & noise correlations for familiar vs. novel images
    - Influence of choice on signal & noise correlations

Exercises: 
* Decoding choice - how quickly can you decode choice, from different brain areas, etc

**Comment by matchings** [comment]
6/28/24 meeting notes: 
* importance of choosing good examples (cells & sessions) that actually show an effect, to prevent confusion 
* last year students got lost because example cells didn't show the effects we were trying to demonstrate
* using more than one cell is helpful to prevent boredom & show generalizeability of effects
* framing the problem around question, with expectations of what results to see (ex: Vip cells are running modulated, we expect to see that)
* avoid situations where we ask the students to do the exact same thing that they were just shown how to do (ex: do the same thing with another cell / session) - needs to be something new, but not so new that its too difficult
* one way to do this is to have them do comparisons (ex: what happens if you try this for active vs. passive, or whatever other condition) that use similar methods but to ask a different kind of question
* another example - do this for one cell, then do it for all the cells and look at variability / patterns across the population 
* challenge of population decoding - it can be too good! helps to make the time bins smaller, allows you to look at timecourse of decoding & understanding of how different parameters of the analysis affect the results 
* can also try decoding from different #s of neurons; this can give insight into the role of noise corerlations

Plan: 
* Meet again Monday
* Marina & Gabe will have full draft of notebooks (i.e. runs all the way through, maybe missing explanatory detail)
* Pass to TAs for review & elaboration before Tuesday meeting
* Matt & Yoni will help with workshop 1 
* Johannes will help with workshop 2 
* Then we will swap
